feat(theme): add contrast opacity values for all color types and hues

Splaktar opened the pull request:
<!-- 
Filling out this template is required! Do not delete it when submitting a Pull Request! Without this information, your Pull Request may be auto-closed.
-->
## PR Checklist
Please check that your PR fulfills the following requirements:
- [x] The commit message follows [our guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format)
- [x] Tests for the changes have been added or this is not a bug fix / enhancement
- [x] Docs have been added, updated, or were not required

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Enhancement
[ ] Documentation content changes
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying and link to one or more relevant issues. -->
Issue Number: 
N/A

## What is the new behavior?

- Add opacity keyword support (`secondary`, `icon`, `disabled`, `hint`, `divider`)
- Deprecate documentation of `foreground-*` in favor of `background-default-contrast-*`
- Allow `foregroundPalette` to override colors on default background and hues of equal contrast type

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
<!-- Note that breaking changes are highly unlikely to get merged to master unless the validation is clear and the use case is critical. -->

## Other information
This is based on @clshortfuse's PR https://github.com/angular/material/pull/8872.

Splaktar commented:
TODO
- [x] Test basic features of md-colors
- [ ] Test dark modes
- [ ] Test demos of all components

Splaktar commented:
This PR alone doesn't fix #7987 but I've addressed that fix in PR #11378 which is layered on top of this branch.

Splaktar commented:
PR https://github.com/angular/material/pull/11380 open for Bottom Sheet dark theme issues.

Splaktar commented:
Opened PR https://github.com/angular/material/pull/11381 to address the various dark mode demo issues which don't appear to be related to this PR at all.

clshortfuse commented:
Looking good.

I don't have the time to review all these changes, but good to see my work not going to waste.

Splaktar commented:
This PR has about 42 screenshot test failures which still need to be evaluated to see if they are acceptable or not.

josephperrott commented:
Confirmed CLA manually.

clshortfuse commented on src/core/services/theming/theming.js:
```diff
@@ -190,6 +190,7 @@ var DARK_DEFAULT_HUES = {
   }
 };
 
+// TODO these references no longer exist with the MD spec update of 2018.
```
Generally, you can replace all `https://material.google.com/` URLs with `https://material.io/archive/guidelines/`

The new URLs would be:

https://material.io/archive/guidelines/style/color.html#color-usability (new fragment) 
https://material.io/archive/guidelines/style/icons.html#icons-system-icons

Splaktar commented:
Rebased and updated.

TODO:
- [ ] Review how this could break existing apps and provide a BREAKING CHANGE: note in commit message if necessary
- [ ] Determine if the Theming Guides need updates after these changes and whether that should be done in a separate PR

Splaktar commented:
I don't believe that the changes to the internal mini-DSL for defining theme files in components should cause any breakages to apps built with the library.

There may be some impacts to `angular/material-tools` and we'll likely need to do a new release for it when we ship `1.2.0`. This is tracked in https://github.com/angular/material-tools/issues/46.

Splaktar commented on src/core/services/theming/theming.js:
```diff
@@ -190,10 +207,8 @@ var DARK_DEFAULT_HUES = {
-// not inactive icon opacity from https://material.google.com/style/icons.html#icons-system-icons
-
+// Icon opacity values (active/inactive) from
+// https://material.io/archive/guidelines/style/icons.html#icons-system-icons
```
You have to scroll down a bit, but yeah that looks like a better link. Thanks!
